drivers: mipi_dsi: implement clock selection algorithm

danieldegrasse opened the pull request:
With the phy-clock being specified in devicetree (and thus under user
control), there is no need to artificially enlarge the DPHY clock to
insure it is fast enough. Instead, we can calculate the DPHY clock
directly, selecting the closest realizable value that is at least as
fast as the value requested by the user.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/59215

danieldegrasse committed:
drivers: mipi_dsi: implement clock selection algorithm

With the phy-clock being specified in devicetree (and thus under user
control), there is no need to artificially enlarge the DPHY clock to
insure it is fast enough. Instead, we can calculate the DPHY clock
directly, selecting the closest realizable value that is at least as
fast as the value requested by the user.

Fixes #59215

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

decsny commented on drivers/mipi_dsi/dsi_mcux.c:
```diff
@@ -23,10 +23,12 @@ LOG_MODULE_REGISTER(dsi_mcux, CONFIG_MIPI_DSI_LOG_LEVEL);
  *         (Pixel clock * bit per output pixel) / number of MIPI data lane
  *
- * Here the desired DPHY bit clock multiplied by ( 9 / 8 = 1.125) to ensure
+ * Here the desired DPHY bit clock multiplied by ( 18 / 17 = 1.058) to ensure
```
magic value

danieldegrasse replied:
It is *unquestionably* a magic value. The issue that this clock enlargement solves is that the MIPI DSI clock setup function `DSI_DphyGetPllDivider` (in the SDK, `drivers/mipi_dsi/fsl_mipi_dsi.c`) calculates the "best" dividers for the MIPI DPHY bit clock given a target frequency. **However**, the DPHY bit clock must be at least `(pixel clock * bits per output pixel) / number of mipi data lanes` (as the comment suggests). The clock enlargement is run because we need to be sure that the actual output frequency is at least this value, and the `DSI_DphyGetPllDivider function will use a lower frequency than the requested value if that is the closest frequency to what is requested that the hardware can actually produce.

The issue that arose here was that previously we calculated the MIPI DPHY like so:
`(LCDIF clock frequency *  bits per output pixel) / number of data lanes`. For the RK055HDMIPI4M, the clock enlargement of 1.125x did not cause issues. Now that we specific the MIPI DPHY bit clock directly in devicetree, the enlargement process violates a value in the reference manual: the MIPI DPHY bit clock should not exceed 800MHz. One solution to this would be to keep the existing clock enlargement, and simply cap the max value the calculated DPHY bit clock can be at 800MHz. The reason I also reduced the aggressiveness of the DPHY bit clock enlargement is because the user now can control the DPHY clock frequency more directly via the devicetree property, so if the enlargement is not sufficient they can just increase their clock value in devicetree.

I arrived at (18/17) because when applied to the DPHY bit clock frequency we place in the devicetree for the RK055HDMIPI4M display (the only MIPI display we currently support upstream with this driver), that results in a round 792MHz DPHY bit clock being requested from the MIPI. Another option we have here is to remove the clock enlargement entirely, since we can modify the value via devicetree now. I think the utility it provides now is a heuristic to attempt to avoid the SDK's decision to select the "best fit" MIPI DPHY clock, rather than biasing towards the higher frequency one (as it should, IMHO)

decsny replied:
It sounds like the SDK is doing the right thing by selecting the closest value, this is just a convenience for portability. I think we should remove this clock enlargement because I don't see the value in it and it just looks like it adds an ounce of confusion. 

If a user puts an invalid value for their hardware in devicetree, it is not a driver's responsibility to correctly interpret what they "meant". If devicetree is correctly used and the datasheets are correctly read then it shouldn't matter what the SDK's estimation behavior is because the value should be an exact match.

decsny replied:
The point of devicetree is to describe hardware, if you change your hardware you need to say what hardware you changed to, ideally if these values are based on the SOC or display hardware then there will already be a dtsi describing them which can be consumed by the driver. Maybe I am not understanding the issue here.

decsny replied:
To me from what you're saying, it sounds like the DPHY reference clock frequency is a board specific property, not specific to the MIPI or SOC alone, is that incorrect?

danieldegrasse replied:
> The point of devicetree is to describe hardware

Indeed, and the devicetree for the display will describe the display. The PHY clock required is a function of the number of the pixels on the display, the bit count of each pixel, and the framerate. It has no grounding in what SOC drives the display, unless you SOC cannot drive the display at the configured framerate.

If you change your SOC, you have not changed your display, and in my view we should make any display definition as portable between SOCs as possible. That means that if the display requests a given PHY clock, the resulting PHY clock should ideally be the same or faster than the requested one, so that the display works correctly on the new SOC without editing your display's devicetree definition.

danieldegrasse commented:
@decsny I have updated this pull request to no longer include the clock enlargement code- as you suggested offline, I have added a clock calculation algorithm to the DSI driver, which will select the closest realizable DPHY clock that is at least as fast as the requested clock. This way users will be able to specify exact values for the DPHY clock if they desire, and display shields that specify a more generic minimum supported DPHY clock will still work.

danieldegrasse commented:
@decsny Ping- can you revisit?
